clippy_lints/src/literal_string_with_formatting_arg.rs

```diff
@@ -0,0 +1,93 @@
+use rustc_ast::ast::{Expr, ExprKind};
+use rustc_ast::token::LitKind;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::BytePos;
+
+use clippy_utils::diagnostics::span_lint;
+use regex::Regex;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks if string literals have formatting arguments outside of macros
+    /// using them (like `format!`).
+    ///
+    /// ### Why is this bad?
+    /// It will likely not generate the expected content.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let x: Option<usize> = None;
+    /// let y = "hello";
+    /// x.expect("{y:?}");
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let x: Option<usize> = None;
+    /// let y = "hello";
+    /// x.expect(&format!("{y:?}"));
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub LITERAL_STRING_WITH_FORMATTING_ARG,
+    suspicious,
+    "Checks if string literals have formatting arguments"
+}
+
+pub struct LiteralStringWithFormattingArg {
+    regex: Regex,
+}
+
+impl LiteralStringWithFormattingArg {
+    pub fn new() -> Self {
+        Self {
+            regex: Regex::new(r"([^\{]\{\w+:?[^\}]*\})|([^\{]\{})|(^\{\w+:?[^\}]*\})|(^\{})").unwrap(),
```

Comment:
This regex is really horrible, if someone has a better idea, I'm interested!

Reply:
I suggest you at least use verbose mode so you can split the regex up and add comments to it:

https://docs.rs/regex/latest/regex/#example-verbose-mode

Reply:
Good idea, thanks!